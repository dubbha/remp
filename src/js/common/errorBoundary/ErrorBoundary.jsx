import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './style.sass';

export default class ErrorBoundary extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      hasError: false,
      error: null,
      info: null,
      detailsHidden: true,
    };
  }

  componentDidCatch(error, info) {
    this.setState({
      hasError: true,
      error,
      info,
    });
  }

  handleDetailsClick = () => {
    this.setState({ detailsHidden: !this.state.detailsHidden });
  }

  render() {
    if (this.state.hasError) {
      return (
        <div className="errorBoundary">
          <div>
            Error [
            <a
              role="link"
              tabIndex={0}
              className="errorBoundary__link"
              onClick={this.handleDetailsClick}
            >
              details
            </a>
            ]
          </div>
          <div
            className={classNames(
              'errorBoundary__details',
              { errorBoundary__details_hidden: this.state.detailsHidden },
            )}
          >
            <div>{this.state.error.toString()}</div>
            <div>{this.state.info.componentStack.split('\n').map(item => (<div key={item}>{item}</div>))}</div>
          </div>
        </div>
      );
    }
    return this.props.children;
  }
}
