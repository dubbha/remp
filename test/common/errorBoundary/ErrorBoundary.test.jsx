import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ErrorBoundary from 'common/errorBoundary';

describe('ErrorBoudnary', () => {
  it('should render children successfully by default', () => {
    const wrapper = shallow(<ErrorBoundary><div>children</div></ErrorBoundary>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update local state hasError flag on componentDidCatch', () => {
    const wrapper = shallow(<ErrorBoundary><div>children</div></ErrorBoundary>);

    wrapper.instance().componentDidCatch('ERROR', { componentStack: 'last\nmid\nfirst' });

    expect(wrapper.state('hasError')).toBe(true);
  });

  it('should update local state error on componentDidCatch', () => {
    const wrapper = shallow(<ErrorBoundary><div>children</div></ErrorBoundary>);

    wrapper.instance().componentDidCatch('ERROR', { componentStack: 'last\nmid\nfirst' });

    expect(wrapper.state('error')).toBe('ERROR');
  });

  it('should update local state info on componentDidCatch', () => {
    const wrapper = shallow(<ErrorBoundary><div>children</div></ErrorBoundary>);

    wrapper.instance().componentDidCatch('ERROR', { componentStack: 'last\nmid\nfirst' });

    expect(wrapper.state('info')).toEqual({ componentStack: 'last\nmid\nfirst' });
  });

  it('should render error block when error caught', () => {
    const wrapper = shallow(<ErrorBoundary><div>children</div></ErrorBoundary>);

    wrapper.setState({
      hasError: true,
      error: 'ERROR',
      info: { componentStack: 'last\nmid\nfirst' },
    });

    expect(toJson(wrapper.render())).toMatchSnapshot();
  });

  it('should expand error stack details on click', () => {
    const wrapper = shallow(<ErrorBoundary><div>children</div></ErrorBoundary>);

    wrapper.setState({
      hasError: true,
      error: 'ERROR',
      info: { componentStack: 'last\nmid\nfirst' },
    });

    wrapper.find('a').simulate('click');

    expect(wrapper.state('detailsHidden')).toBe(false);
  });
});
