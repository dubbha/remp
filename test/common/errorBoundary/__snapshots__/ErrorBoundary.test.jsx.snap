// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoudnary should render children successfully by default 1`] = `
<div>
  children
</div>
`;

exports[`ErrorBoudnary should render error block when error caught 1`] = `
<div
  class="errorBoundary"
>
  <div>
    Error [
    <a
      class="errorBoundary__link"
      role="link"
      tabindex="0"
    >
      details
    </a>
    ]
  </div>
  <div
    class="errorBoundary__details errorBoundary__details_hidden"
  >
    <div>
      ERROR
    </div>
    <div>
      <div>
        last
      </div>
      <div>
        mid
      </div>
      <div>
        first
      </div>
    </div>
  </div>
</div>
`;
